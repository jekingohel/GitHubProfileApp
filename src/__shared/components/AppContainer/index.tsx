import React, {ReactNode} from 'react';
import {
  KeyboardAvoidingView,
  Platform,
  SafeAreaView,
  ScrollView,
  StatusBar,
} from 'react-native';

interface AppContainerProps {
  scrollBar?: boolean;
  scroll?: boolean;
  statusBarColor?: string | null;
  loader?: boolean;
  loaderSize?: number;
  children: ReactNode;
}

/**
 * Functional component for a container that handles keyboard avoiding and status bar configuration.
 * @param {object} props - Component properties.
 * @param {boolean} [props.scrollBar=true] - Controls the visibility of the vertical scroll bar in the ScrollView.
 * @param {boolean} [props.scroll=true] - Enables or disables scrolling behavior.
 * @param {string | null} [props.statusBarColor='#FFFFFF'] - Color of the status bar.
 * @param {boolean} [props.loader=false] - Flag indicating whether a loader is displayed.
 * @param {number} [props.loaderSize] - Size of the loader.
 * @param {ReactNode} props.children - The content to be rendered inside the container.
 * @returns {JSX.Element} - AppContainer component.
 * @example
 * <AppContainer
 *   scrollBar={true}
 *   scroll={true}
 *   statusBarColor="#FFFFFF"
 *   loader={false}
 *   loaderSize={30}
 * >
 *   // Content goes here
 * </AppContainer>
 */
const AppContainer: React.FC<AppContainerProps> = props => {
  const {scrollBar, scroll = true} = props;
  const keyboardVerticalOffset = Platform.OS === 'ios' ? 40 : 0;

  return (
    <SafeAreaView style={{flex: 1}}>
      {/* KeyboardAvoidingView provides automatic keyboard behavior */}
      <KeyboardAvoidingView
        style={{flex: 1, backgroundColor: '#FFFFFF'}}
        behavior={Platform.OS === 'ios' ? 'padding' : undefined}
        keyboardVerticalOffset={keyboardVerticalOffset}
        {...props}>
        {/* StatusBar for controlling status bar appearance */}
        <StatusBar
          animated={false}
          backgroundColor={'#FFFFFF'}
          hidden={false}
        />

        {scroll ? (
          // ScrollView provides a scrollable container for content
          <ScrollView
            keyboardShouldPersistTaps="handled"
            showsVerticalScrollIndicator={scrollBar ?? false}
            contentContainerStyle={{flex: 1}}
            {...props}>
            {props.children}
          </ScrollView>
        ) : (
          // Render content directly if scrolling is disabled
          props.children
        )}
      </KeyboardAvoidingView>
    </SafeAreaView>
  );
};

export default AppContainer;
