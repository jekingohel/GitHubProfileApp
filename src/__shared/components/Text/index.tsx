import React from 'react';
import {Text as RNText, TextStyle} from 'react-native';

type FontWeight =
  | 'normal'
  | 'bold'
  | '100'
  | '200'
  | '300'
  | '400'
  | '500'
  | '600'
  | '700'
  | '800'
  | '900'
  | undefined;

type Align = 'left' | 'auto' | 'right' | 'center' | 'justify' | undefined;

interface CustomViewProps {
  children: React.ReactNode;
  fontSize?: number;
  fontWeight?: FontWeight;
  align?: Align;
  color?: string;
  italic?: boolean;
}

const Text: React.FC<CustomViewProps> = props => {
  const {
    fontSize = 12,
    fontWeight = '400',
    align = 'left',
    color = 'black',
    italic = false,
  } = props;

  const containerStyle: TextStyle = {
    fontSize,
    fontWeight,
    textAlign: align,
    color,
    fontStyle: italic ? 'italic' : 'normal',
  };

  return <RNText style={containerStyle}>{props.children}</RNText>;
};

export default Text;
