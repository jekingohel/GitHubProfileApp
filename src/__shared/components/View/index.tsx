import React from 'react';
import {
  View as RNView,
  ViewStyle,
  ViewProps as RNViewProps,
} from 'react-native';

interface CustomViewProps extends RNViewProps {
  transparent?: boolean;
  alignItems?: ViewStyle['alignItems'];
  justifyContent?: ViewStyle['justifyContent'];
  flexDirection?: ViewStyle['flexDirection'];
  children: React.ReactNode;
  style?: ViewStyle;
}

const View: React.FC<CustomViewProps> = props => {
  const {
    children,
    alignItems = 'stretch',
    justifyContent = 'flex-start',
    flexDirection = 'column',
    ...restProps
  } = props;

  const containerStyle: ViewStyle = {
    backgroundColor: 'transparent',
    alignItems,
    justifyContent,
    flexDirection,
    ...props.style,
  };

  return (
    <RNView testID="test-view" style={containerStyle}>
      {children}
    </RNView>
  );
};

export default View;
