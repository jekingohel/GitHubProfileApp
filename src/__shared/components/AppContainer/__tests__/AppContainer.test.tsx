import React from 'react';
import AppContainer from '..';
import Text from '../../Text';
import View from '../../View';
import renderer from 'react-test-renderer';

test('renders correctly', () => {
  const tree = renderer
    .create(
      <AppContainer>
        <Text>Test Content</Text>
      </AppContainer>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders correctly with scrollable content', () => {
  const tree = renderer
    .create(
      <AppContainer scroll>
        <View style={{height: 900}}>
          <Text>Test Content</Text>
          <Text>Additional Content</Text>
          <Text>More Content</Text>
        </View>
      </AppContainer>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders correctly with a specified status bar color', () => {
  const tree = renderer
    .create(
      <AppContainer statusBarColor="#ff0000">
        <Text>Test Content</Text>
      </AppContainer>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders correctly with a loader', () => {
  const tree = renderer
    .create(
      <AppContainer loader>
        <Text>Test Content</Text>
      </AppContainer>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
