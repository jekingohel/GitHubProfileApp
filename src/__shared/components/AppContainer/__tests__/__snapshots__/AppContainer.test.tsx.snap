// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flex": 1,
        }
      }
      keyboardShouldPersistTaps="handled"
      showsVerticalScrollIndicator={false}
    >
      <View>
        <Text
          style={
            {
              "color": "#000000",
              "fontSize": 12,
              "fontStyle": "normal",
              "fontWeight": "400",
              "lineHeight": 18,
              "textAlign": "left",
            }
          }
        >
          Test Content
        </Text>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;

exports[`renders correctly with a loader 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    loader={true}
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flex": 1,
        }
      }
      keyboardShouldPersistTaps="handled"
      loader={true}
      showsVerticalScrollIndicator={false}
    >
      <View>
        <Text
          style={
            {
              "color": "#000000",
              "fontSize": 12,
              "fontStyle": "normal",
              "fontWeight": "400",
              "lineHeight": 18,
              "textAlign": "left",
            }
          }
        >
          Test Content
        </Text>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;

exports[`renders correctly with a specified status bar color 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    statusBarColor="#ff0000"
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flex": 1,
        }
      }
      keyboardShouldPersistTaps="handled"
      showsVerticalScrollIndicator={false}
      statusBarColor="#ff0000"
    >
      <View>
        <Text
          style={
            {
              "color": "#000000",
              "fontSize": 12,
              "fontStyle": "normal",
              "fontWeight": "400",
              "lineHeight": 18,
              "textAlign": "left",
            }
          }
        >
          Test Content
        </Text>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;

exports[`renders correctly with scrollable content 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    scroll={true}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flex": 1,
        }
      }
      keyboardShouldPersistTaps="handled"
      scroll={true}
      showsVerticalScrollIndicator={false}
    >
      <View>
        <View
          style={
            {
              "alignItems": "stretch",
              "backgroundColor": "transparent",
              "flexDirection": "column",
              "height": 900,
              "justifyContent": "flex-start",
            }
          }
          testID="test-view"
        >
          <Text
            style={
              {
                "color": "#000000",
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "400",
                "lineHeight": 18,
                "textAlign": "left",
              }
            }
          >
            Test Content
          </Text>
          <Text
            style={
              {
                "color": "#000000",
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "400",
                "lineHeight": 18,
                "textAlign": "left",
              }
            }
          >
            Additional Content
          </Text>
          <Text
            style={
              {
                "color": "#000000",
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "400",
                "lineHeight": 18,
                "textAlign": "left",
              }
            }
          >
            More Content
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
